tests/test_cleaneval.py

```diff
@@ -0,0 +1,486 @@
+    else:
+        text = re_TAG.sub(r'\n\g<1> ', text)  # only break lines before segment markers
+        text = re_WS.sub(" ", text)           # normalise whitespace (including line breaks) to single spaces
+        if ascii:
+            text = re_HI.sub("", text)        # delete non-ASCII characters (to avoid charset problems)
+        return text
+
+## return diff as list of tuples ("equal"/"insert"/"delete", [text words], [gold words])
+def make_diff(alignment, text_w, gold_w):
+	diff = []
+	for tag, i1, i2, j1, j2 in alignment.get_opcodes():
+		text_region = text_w[i1:i2]
+		gold_region = gold_w[j1:j2]
+		if tag == "replace":
+			diff.append( ("delete", text_region, []) )
+			diff.append( ("insert", [], gold_region) )
+		else:
+			diff.append( (tag, text_region, gold_region) )
+	return diff
+
+## return evaluation measures for given diff:
+##   (f-score, precision, recall, [labelled] segmentation f-score, precision, recall)
+def evaluate(diff):
+    tp = fp = fn = 0
+    tag_tp = tag_fp = tag_fn = 0
+    for tag, text, gold in diff:
+        text_tags = len(list(filter(re_TAG.match, text)))
+        gold_tags = len(list(filter(re_TAG.match, gold)))
+        text_l = len(text)
+        gold_l = len(gold)
+        if tag == "delete":
+            fp += text_l
+            tag_fp += text_tags
+        elif tag == "insert":
+            fn += gold_l
+            tag_fn += gold_tags
+        else:
+            tp += text_l
+            tag_tp += text_tags
+            assert text_l == gold_l
+            assert text_tags == gold_tags
+    result = __evaluate(tp, fp, fn, tag_tp, tag_fp, tag_fn)
+    return result
+
+def __evaluate(tp, fp, fn, tag_tp, tag_fp, tag_fn):
+    n_text = tp + fp if tp + fp > 0 else 1                                                                                                              
+    n_gold = tp + fn if tp + fn > 0 else 1
+    precision = float(tp) / n_text
+    recall = float(tp) / n_gold
+    precision_plus_recall = precision + recall if precision + recall > 0 else 1
+    f_score = 2 * precision * recall / precision_plus_recall
+                                                                                                                                                        
+    tags_text = tag_tp + tag_fp if tag_tp + tag_fp > 0 else 1
+    tags_gold = tag_tp + tag_fn if tag_tp + tag_fn > 0 else 1
+    tag_precision = float(tag_tp) / tags_text
+    tag_recall = float(tag_tp) / tags_gold
+    precision_plus_recall = tag_precision + tag_recall if tag_precision + tag_recall > 0 else 1
+    tag_f_score = 2 * tag_precision * tag_recall / precision_plus_recall
+                                                                                                                                                        
+    return (100 * f_score, 100 * precision, 100 * recall, 100 * tag_f_score, 100 * tag_precision, 100 * tag_recall, tp, fp, fn, tag_tp, tag_fp, tag_fn)
+
+def write_alignment(diff, filename):
```

Comment:
what is the purpose of this function?

every function needs a docstring explaining this

Reply:
a quick search of your code makes it look like this function isn't even getting called, in which case you should just delete it

we call code that's not being used "dead code", and it is a real maintenance burden